src/ray/raylet/local_placement_group_manager.h

```diff
@@ -0,0 +1,131 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/bundle_spec.h"
+#include "ray/common/id.h"
+#include "ray/common/task/scheduling_resources.h"
+
+namespace ray {
+
+namespace raylet {
+
+enum CommitState {
+  /// Resources are prepared.
+  PREPARED,
+  /// Resources are COMMITTED.
+  COMMITTED
+};
+
+struct BundleState {
+  /// Leasing state for 2PC protocol.
+  CommitState state;
+  /// Resources that are acquired at preparation stage.
+  ResourceIdSet acquired_resources;
+};
+
+struct pair_hash {
+  template <class T1, class T2>
+  std::size_t operator()(const std::pair<T1, T2> &pair) const {
+    return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+  }
+};
+
+/// LocalPlacementGroupManagerInterface is responsible for handle the register bundle
+/// request from GCS.
+class LocalPlacementGroupManagerInterface {
```

Comment:
Can we update these docs to include some more detailed information?

For example, `PrepareBundle` is responsible for allocating/reserving the necessary resources, `CommitBundle` is responsible for creating new custom resources, `ReturnBundle` is _not_ responsible for killing workers, etc.

Reply:
sure, thank u.

Reply:
Do we already documented the 2PC protocol somewhere? If so, we can document something like this here:
```
`LocalPlacementGroupManagerInterface` responsible for managing the resources that about allocated for placement group bundles. See ... for the details for the 2PC protocol of bundle assignment between GCS and Raylet.
```

Reply:
Also, about naming, I'd name it `PlacementGroupResourceManager`. Because 1) "local" and "interface" are not informative here. 2) This class is for managing resources, not place groups themselves.

Reply:
Good idea, has rename it to `PlacementGroupResourceManager `